Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp

```diff
@@ -0,0 +1,438 @@
+{
+    ASSERT(checkRequirements());
+    return std::unique_ptr<AcceleratedBackingStoreDMABuf>(new AcceleratedBackingStoreDMABuf(webPage));
+}
+
+AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf(WebPageProxy& webPage)
+    : AcceleratedBackingStore(webPage)
+{
+}
+
+AcceleratedBackingStoreDMABuf::~AcceleratedBackingStoreDMABuf()
+{
+    if (m_surface.id)
+        m_webPage.process().removeMessageReceiver(Messages::AcceleratedBackingStoreDMABuf::messageReceiverName(), m_surface.id);
+
+    if (m_gdkGLContext && m_gdkGLContext.get() == gdk_gl_context_get_current())
+        gdk_gl_context_clear_current();
+}
+
+AcceleratedBackingStoreDMABuf::RenderSource::RenderSource(const WebCore::IntSize& size, float deviceScaleFactor)
+    : m_size(size)
+    , m_deviceScaleFactor(deviceScaleFactor)
+{
+}
+
+AcceleratedBackingStoreDMABuf::Texture::Texture(GdkGLContext* glContext, const UnixFileDescriptor& backFD, const UnixFileDescriptor& frontFD, int fourcc, int32_t offset, int32_t stride, const WebCore::IntSize& size, float deviceScaleFactor)
+    : RenderSource(size, deviceScaleFactor)
+    , m_context(glContext)
+{
+    gdk_gl_context_make_current(m_context.get());
+    auto& display = WebCore::PlatformDisplay::sharedDisplay();
+    Vector<EGLAttrib> attributeList = {
+        EGL_WIDTH, m_size.width(),
+        EGL_HEIGHT, m_size.height(),
+        EGL_LINUX_DRM_FOURCC_EXT, fourcc,
+        EGL_DMA_BUF_PLANE0_FD_EXT, backFD.value(),
+        EGL_DMA_BUF_PLANE0_OFFSET_EXT, offset,
+        EGL_DMA_BUF_PLANE0_PITCH_EXT, stride,
+        EGL_NONE };
+    m_backImage = display.createEGLImage(EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, nullptr, attributeList);
+    if (!m_backImage) {
+        WTFLogAlways("Failed to create EGL image from DMABuf with file descriptor %d", backFD.value());
+        return;
+    }
+    attributeList[7] = frontFD.value();
+    m_frontImage = display.createEGLImage(EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, nullptr, attributeList);
+    if (!m_frontImage) {
+        WTFLogAlways("Failed to create EGL image from DMABuf with file descriptor %d", frontFD.value());
+        display.destroyEGLImage(m_backImage);
+        m_backImage = nullptr;
+        return;
+    }
+
+    glGenTextures(1, &m_textureID);
+    glBindTexture(GL_TEXTURE_2D, m_textureID);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+#if USE(GTK4)
+    glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_backImage);
+    m_texture[0] = adoptGRef(gdk_gl_texture_new(m_context.get(), m_textureID, m_size.width(), m_size.height(), nullptr, nullptr));
+    m_texture[1] = adoptGRef(gdk_gl_texture_new(m_context.get(), m_textureID, m_size.width(), m_size.height(), nullptr, nullptr));
```

Comment:
Because of the way gdk works internally, if you pass a different GdkTexture object even if for the same texture, it makes a diff to get the changed region. That's whay I understood, see https://bugs.webkit.org/show_bug.cgi?id=252648